database/modules/user.model.js

```diff
@@ -0,0 +1,62 @@
+const mongoose = require("mongoose");
+const bcrypt = require("bcrypt");
+const joi = require("joi");
+const jsonWebToken = require("jsonwebtoken");
+const config = require("../../config.env");
+const { Timestamp, ObjectId } = require("mongodb");
+
+const UserSchema = new mongoose.Schema(
+  {
+    name: { type: String, required: true },
+    email: {
+      type: String,
+      unique: true,
+      validate: {
+        validator(email) {
+          const { error } = joi.string().email().validate(email);
+
+          if (error) throw new Error("Email not vlid");
+        },
+      },
+    },
+    passwordHash: { type: String },
+    familyId: { type: ObjectId, default: null },
+    tokens: [
+      {
+        token: { type: String, require: true },
+        expires: { type: Date, require: true },
+      },
+    ],
+    verificationToken: { type: String },
+  },
+  { timestamps: true }
+);
+
+UserSchema.static("hashPassword", (password) => {
+  const costFactor = 6;
+  return bcrypt.hash(password, costFactor);
+});
+
+UserSchema.method("isPasswordValid", function (password) {
+  return password.compare(password, this.passwordHash);
+});
+
+UserSchema.method("generateAndSaveToken", async function () {
+  const token = jsonWebToken.sign({ id: this._id }, config.jwtPrivateKey);
+
+  this.tokens.push({
+    token,
+    expires: new Date().getTime() + 24 * 60 * 60 * 1e3,
+  });
+  await this.save();
+
+  return token;
+});
+
+UserSchema.pre("save", async function () {
```

Comment:
Since we're going to implement OAuth2.0 (we won't save password in db). I don't think that it's a good idea to always hash password upon saving user. Especially with some weird `isNew` flag